### ISSUE TITLE
Pretrained model

### ISSUE THREAD

**Issue by adityac95** [Tsung-Ping/Harmony-Transformer-v2#1]
Hi,
Do you have a pretrained version of this model available? I'd like to use it to predict chords in an unlabelled MIDI corpus but training the model from scratch is a big time obstacle.
Thanks
Aditya

**Comment by Tsung-Ping** [comment]
Hi Aditya,
I am sorry I did not save a pretrained model. I think it would not take a long time to train the model (no longer than one day with GPU).

**Comment by napulen** [comment]
I managed to train it with a `tensorflow:1.15.5-gpu-py3` docker container.

Here is the `Dockerfile`:

```Dockerfile
FROM tensorflow/tensorflow:1.15.5-gpu-py3

# https://github.com/NVIDIA/nvidia-docker/issues/1631#issuecomment-1112682423
RUN rm /etc/apt/sources.list.d/cuda.list
RUN rm /etc/apt/sources.list.d/nvidia-ml.list
RUN apt-key del 7fa2af80
RUN apt-get update && apt-get install -y --no-install-recommends wget
RUN wget https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64/cuda-keyring_1.0-1_all.deb
RUN dpkg -i cuda-keyring_1.0-1_all.deb

COPY . /HT2
WORKDIR /HT2
RUN apt update
# RUN apt install -y python3-tk
RUN python -m pip install -r requirements.txt
```

Then you 
```bash
docker build . -t chensu21
docker run -it chensu21

# Inside the container
python BPS_FH_preprocessing.py
python functional_harmony_recognition.py
```

**Comment by napulen** [comment]
Here is the pretrained model, it ran for ~40 epochs before the early stopping kicked in.

[model.zip](https://github.com/Tsung-Ping/Harmony-Transformer-v2/files/9263619/model.zip)

**Comment by napulen** [comment]
A very different problem is how to run it for inference on a new dataset. I haven't been able to.

@Tsung-Ping if you have time to meet with me this week to teach me how to use the model for inference, it'd mean the world to me. Thanks.

**Comment by Tsung-Ping** [comment]
Hi, Néstor

When inferencing on new data, you have to represent the data according to the input formats defined in 'placeholder' (see  train_HT ()):
>     with tf.name_scope('placeholder'):
>         x_p = tf.placeholder(tf.int32, [None, hp.n_steps, 88], name="pianoroll")
>         x_len = tf.placeholder(tf.int32, [None], name="seq_lens")
>         slope = tf.placeholder(dtype=tf.float32, name='annealing_slope')
where
*x_p* is the piano-roll representation of a midi file, which has the shape of [batch size, time steps, 88 midi numbers],
*x_len* is the 'valid' lengths (the length without paddings) of the batched pianorolls,
*slope* is a fixed number based on your training result (it should be a float number larger than 1).

After the preprocessing, you have to restore your model (https://www.tensorflow.org/api_docs/python/tf/compat/v1/train/Saver#restore) and write a few lines of codes as shown below  (also in  train_HT ()) to feed the data into the model and inference the chords:

> # Testing
> test_run_list = [pred_k, pred_r] # for inference, only 'key' and 'Roman numeral' are needed
> test_feed_fict = {x_p: test_data['pianoroll'], # the data you want to inference on
>                            x_len: test_data['len'], # valid lengths
>                            y_tc: test_data['tchord'], # ground truth for evaluation; not needed during inference
>                            dropout: 0.0, # set to 0 during inference
>                            is_training: False}  # set to False during inference
> test_pred_k, test_pred_r = sess.run(test_run_list, feed_dict=test_feed_fict)
The prediction results would be test_pred_k (key) and test_pred_r  (Roman numeral).

**Comment by napulen** [comment]
Thanks for the reply! I'll try this out.

**Comment by napulen** [comment]
@Tsung-Ping thank you for the help! I was able to make it work. Here is my resulting inference function:

```python
def inference_HT(model_checkpoint, annealing_slope=1.0):
    print('Run HT functional for inference: %s...' % (model_checkpoint))
    _, test_data = load_data_functional(dir=hp.dataset + '_preprocessed_data_MIREX_Mm.pickle', test_set_id=hp.test_set_id, sequence_with_overlap=hp.train_sequence_with_overlap)

    with tf.name_scope('placeholder'):
        x_p = tf.placeholder(tf.int32, [None, hp.n_steps, 88], name="pianoroll")
        x_len = tf.placeholder(tf.int32, [None], name="seq_lens")
        dropout = tf.placeholder(dtype=tf.float32, name="dropout_rate")
        is_training = tf.placeholder(dtype=tf.bool, name="is_training")
        slope = tf.placeholder(dtype=tf.float32, name='annealing_slope')

    with tf.name_scope('model'):
        x_in = tf.cast(x_p, tf.float32)
        source_mask = tf.sequence_mask(lengths=x_len, maxlen=hp.n_steps, dtype=tf.float32) # [n_batches, n_steps]
        target_mask = source_mask
        chord_change_logits, dec_input_embed, _, _, _, _ = crm.HTv2(x_in, source_mask, target_mask, slope, dropout, is_training, hp)

    with tf.variable_scope("output_projection"):
        n_key_classes = 42 + 1
        n_roman_classes = 9 * 14 * 10 * 4 + 1
        dec_input_embed = tf.layers.dropout(dec_input_embed, rate=dropout, training=is_training)
        key_logits = tf.layers.dense(dec_input_embed, n_key_classes)
        roman_logits = tf.layers.dense(dec_input_embed, n_roman_classes)

    with tf.name_scope('evaluation'):
        pred_cc = tf.cast(tf.round(tf.sigmoid(slope*chord_change_logits)), tf.int32)
        pred_k = tf.argmax(key_logits, axis=2, output_type=tf.int32)
        pred_r = tf.argmax(roman_logits, axis=2, output_type=tf.int32)

    saver = tf.train.Saver(max_to_keep=1)

    # Inference
    print('Run the model...')
    with tf.Session() as sess:
        saver.restore(sess, model_checkpoint)
        test_run_list = [pred_cc, pred_k, pred_r]
        test_feed_fict = {x_p: test_data['pianoroll'],
                          x_len: test_data['len'],
                          dropout: 0.0,
                          is_training: False,
                          slope: annealing_slope}
        test_pred_cc, test_pred_k, test_pred_r = sess.run(test_run_list, feed_dict=test_feed_fict)
    return test_pred_cc, test_pred_k, test_pred_r
```

**Comment by napulen** [comment]
A follow-up question: how important is the annealing slope? And could I use the annealing slope of one run in a different checkpoint? I trained the model twice, in one instance, I saved the weights but lost the slope; in another run I kept the logs (thus the slope) but lost the weights. Would it make sense to combine the weights of one and the slope of other? This was on the BPS data, so same dataset.

**Comment by Tsung-Ping** [comment]
Hi, Néstor,

You can infer the value of the slope according to the following codes (in train_HT()):
>                if step > 0 and step % n_iterations_per_epoch == 0:
>                    annealing_slope *= hp.annealing_rate

For more details of the slope annealing trick, you may refer to the paper: "Junyoung Chung, Sungjin Ahn, and Yoshua Bengio, Hierarchical multiscale recurrent neural networks, ICLR 2017."

**Comment by Tsung-Ping** [comment]
Hi, Néstor,

You can infer the value of the slope according to the following codes (in
train_HT()):

>             if step > 0 and step % n_iterations_per_epoch == 0:
>                 annealing_slope *= hp.annealing_rate


For more details of the slope annealing trick, you may refer to the paper:
"Junyoung Chung, Sungjin Ahn, and Yoshua Bengio, Hierarchical multiscale
recurrent neural networks, ICLR 2017."

Néstor Nápoles López ***@***.***> 於 2022年8月6日 週六 上午9:57寫道：

**Comment by napulen** [comment]
I just realized I never replied back. This worked, I was able to train the model effectively. Thank you!
